crates/nu-lsp/src/symbols.rs

```diff
@@ -0,0 +1,534 @@
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Symbol {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.kind == other.kind {
+            return self.range.start.cmp(&other.range.start);
+        }
+        match (self.kind, other.kind) {
+            (SymbolKind::FUNCTION, _) => Ordering::Less,
+            (_, SymbolKind::FUNCTION) => Ordering::Greater,
+            _ => self.range.start.cmp(&other.range.start),
+        }
+    }
+}
+
+impl Symbol {
+    fn to_symbol_information(&self, uri: &Uri) -> SymbolInformation {
+        #[allow(deprecated)]
+        SymbolInformation {
+            location: Location {
+                uri: uri.clone(),
+                range: self.range,
+            },
+            name: self.name.to_owned(),
+            kind: self.kind,
+            container_name: None,
+            deprecated: None,
+            tags: None,
+        }
+    }
+}
+
+/// Cache symbols for each opened file to avoid repeated parsing
+pub struct SymbolCache {
+    /// Fuzzy matcher for symbol names
+    matcher: SkimMatcherV2,
+    /// File Uri --> Symbols
+    cache: BTreeMap<Uri, Vec<Symbol>>,
+    /// If marked as dirty, parse on next request
+    dirty_flags: BTreeMap<Uri, bool>,
+}
+
+impl SymbolCache {
+    pub fn new() -> Self {
+        SymbolCache {
+            matcher: SkimMatcherV2::default(),
+            cache: BTreeMap::new(),
+            dirty_flags: BTreeMap::new(),
+        }
+    }
+
+    pub fn mark_dirty(&mut self, uri: Uri, flag: bool) {
+        self.dirty_flags.insert(uri, flag);
+    }
+
+    fn get_symbol_by_id(
+        working_set: &StateWorkingSet,
+        id: Id,
+        doc: &FullTextDocument,
+        doc_span: &Span,
+    ) -> Option<Symbol> {
+        match id {
+            Id::Declaration(decl_id) => {
+                let decl = working_set.get_decl(decl_id);
+                let span = working_set.get_block(decl.block_id()?).span?;
+                // multi-doc working_set, returns None if the Id is in other files
+                if !doc_span.contains(span.start) {
+                    return None;
+                }
+                Some(Symbol {
+                    name: decl.name().to_string(),
+                    kind: SymbolKind::FUNCTION,
+                    range: span_to_range(&span, doc, doc_span.start),
+                })
+            }
+            Id::Variable(var_id) => {
+                let var = working_set.get_variable(var_id);
+                let span = var.declaration_span;
+                if !doc_span.contains(span.start) || span.end == span.start {
+                    return None;
+                }
+                let range = span_to_range(&span, doc, doc_span.start);
+                let name = doc.get_content(Some(range));
+                // TODO: better way to filter closures with type any
+                if name.contains('\r')
+                    || name.contains('\n')
+                    || name.contains('{')
```

Comment:
I don't know if it's supposed to return variables whose span covers an entire block like `{|xx| xxxx}`

Reply:
hmm, good question. we clearly can have those. i wonder what happens with an lsp like rust-analyzer when it has a raw string that covers multiple lines?

I assume you're talking about the span of the contents versus the span of the variable name.

Reply:
I think raw string should be in category value instead of var. I don't see why we give a variable id to anonymous closures in working set, but I guess it's there for some reason.

Reply:
To be clear, the name string here in the variable span, should ideally be the identifier of the variable.

Reply:
oh, you're saying the name is the closure. that seems weird.

Reply:
I wonder if this is returning the same results as `scope variables` because all of those have names.

Reply:
It's more than that, named variables + flags/arguments + anonymous closures + "where predicate" (see my 2nd screenshot, the selected one) + maybe others.

Reply:
I don't know how to provide a minimal reproducible code, but here's what happens if the `contains` checks are removed

<img width="1133" alt="image" src="https://github.com/user-attachments/assets/ab0638f3-79e3-44e5-ae9f-edac7aef1b6b" />

I'm not planning on going deep into the parser to figure out why atm, but let me know if I can provide more info to help debug this.

Reply:
But the problem is that they all come from var Id, and there type is any.

Reply:
I got them from the workingset after parsing the simple script. Maybe those will be removed after merge to engine state, but that's not how we handle the parsing right now in lsp server.

Reply:
I think here's why, the spans for `$in` and `$it` are somewhat larger. However the there's no name in `struct Variable`, hard to identify them with limited info.

https://github.com/nushell/nushell/blob/5cf6dea9976ff969bfeb6c8042181d51a58597ac/crates/nu-parser/src/parser.rs#L3430

https://github.com/nushell/nushell/blob/5cf6dea9976ff969bfeb6c8042181d51a58597ac/crates/nu-parser/src/parser.rs#L6628

Reply:
Interesting. But they're still being named with $it and $in. Why can't they be found by it and in? I could see why the value would be a block with these. 

What happens if you rename them to something else like "row_condition_it" and "in_variable" or some such thing? Does it help to have something more verbose like that or does it just produce the same results?

Reply:
All I get from the lsp side is the var id, then I call `working_set.get_variable` to get the variable span, then read the name according to its span. 

Renaming doesn't help, as the span is unchanged, and will probably breaks the parsing.

One reasonable solution would be, making the span in struct variable to Option type, and None for those with name it/in, if the span for them aren't useful anywhere, which I'm not sure.

Reply:
Let's not change that right now and just land this with how you have it now. I'd like to get this in but also think more about what the "right" thing to do is. What do you think?

Reply:
If your question is why I can't get the variable name, what I can get is the names of the last_overlay.

I think it's not the full list of all variables.

https://github.com/nushell/nushell/blob/79f19f2fc7dcd0da30ec8a35008a26177afccece/crates/nu-protocol/src/engine/state_working_set.rs#L636-L654

Reply:
But I'm not sure you should get the full list of variables. It should only produce the variables in scope I think, right?

Reply:
> Let's not change that right now and just land this with how you have it now. I'd like to get this in but also think more about what the "right" thing to do is. What do you think?

I think that's it.